mender-auth/cli/keystore.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-auth/cli/keystore.hpp>
+
+#include <common/log.hpp>
+#include <common/crypto.hpp>
+
+
+namespace mender {
+namespace auth {
+namespace cli {
+
+using namespace std;
+
+namespace log = mender::common::log;
+
+namespace crypto = mender::common::crypto;
+
+class KeyStoreErrorCategoryClass : public std::error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+
+enum ErrorCode {
+	NoError = 0,
+	NoKeysError,
+	StaticKeyError,
+};
+
+const KeyStoreErrorCategoryClass KeyStoreErrorCategory;
+
+const char *KeyStoreErrorCategoryClass::name() const noexcept {
+	return "KeyStoreErrorCategory";
+}
+
+string KeyStoreErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case NoKeysError:
+		return "No keys found";
+	case StaticKeyError:
+		return "Cannot replace static key";
+	}
+	// Don't use "default" case. This should generate a warning if we ever add any enums. But
+	// still assert here for safety.
+	assert(false);
+	return "Unknown";
+}
+
+const error::Error KeyStoreNoKeysError =
+	error::Error(error_condition(NoKeysError, KeyStoreErrorCategory), "no keys");
```

Comment:
I would recommend not storing const errors like these. The second string field is supposed to provide context for where you are returning the error from, so it should be constructed in-place.

This also means that it's not generally safe to compare the whole error, we generally compare the underlying error_conditions, which (perhaps somewhat confusingly) is stored in the `code` member.

The other files that have defined their own errors make this more convenient by defining `MakeError`, so that you don't have to specify the whole Category boilerplate every time. When comparing you also use `MakeError` to construct a value to compare against (search for `.code`). This last part could arguably be better, but it's the preferred style at the moment.

Reply:
Thank you for the detailed explanation. I'll clean it up.